### PR TITLE
LibTextCodec: Refurbish decoders

### PR DESCRIPTION
**Pull request by AtkinsSJ**
Bring LibTextCodec from the dark ages of `DeprecatedString` and `Decoder*` into the glistening cyberfuture of `ErrorOr<String>` and `Optional<Decoder&>`.

### CODE REVIEW

File: Userland/Libraries/LibWeb/CSS/Parser/Tokenizer.cpp [hunk 1]
```diff
@@ -206,37 +206,39 @@ Tokenizer::Tokenizer(StringView input, StringView encoding)
...
+                   // in input by a single U+000A LINE FEED (LF) code point.
+                   if (code_point == '\r') {
+                       if (last_was_carriage_return) {
+                           builder.append('\n');
+                       } else {
+                           last_was_carriage_return = true;
+                       }
+                   } else {
+                       if (last_was_carriage_return)
+                           builder.append('\n');
```
Comment (ADKaster) [inline Userland/Libraries/LibWeb/CSS/Parser/Tokenizer.cpp:222]: builder.append() in a function returning ErrorOr? :thinking:

File: Meta/Lagom/Fuzzers/FuzzCyrillicDecoder.cpp [hunk 2]
```diff
@@ -11,8 +11,8 @@
 
 extern "C" int LLVMFuzzerTestOneInput(uint8_t const* data, size_t size)
 {
-    auto* decoder = TextCodec::decoder_for("windows-1251"sv);
-    VERIFY(decoder);
-    decoder->to_utf8({ data, size });
+    auto decoder = TextCodec::decoder_for("windows-1251"sv);
+    VERIFY(decoder.has_value());
+    MUST(decoder->to_utf8({ data, size }));
```
Comment (nico) [inline Meta/Lagom/Fuzzers/FuzzCyrillicDecoder.cpp:16]: Doesn't MUST here mean the fuzzer will now find invalid utf-8? That doesn't seem terribly useful 😅

(Also below in the other fuzzers)

File: Tests/LibTextCodec/TestTextDecoders.cpp [hunk 3]
```diff
@@ -15,13 +16,13 @@ TEST_CASE(test_utf8_decode)
...
-        processed_code_points.append(code_point);
-    });
+    MUST(decoder.process(test_string, [&](u32 code_point) {
+        return processed_code_points.try_append(code_point);
+    }));
     EXPECT(processed_code_points.size() == 1);
     EXPECT(processed_code_points[0] == 0x1F600);
 
-    EXPECT(decoder.to_utf8(test_string) == test_string);
+    EXPECT(decoder.to_utf8(test_string).value() == test_string);
```
Comment (nico) [inline Tests/LibTextCodec/TestTextDecoders.cpp:25]: MUST instead of value maybe?

File: Userland/Libraries/LibTextCodec/Decoder.cpp [hunk 4]
```diff
@@ -171,38 +171,43 @@ Decoder* bom_sniff_to_decoder(StringView input)
...
     // 1. Let BOMEncoding be the result of BOM sniffing ioQueue.
     // 2. If BOMEncoding is non-null:
-    if (auto* unicode_decoder = bom_sniff_to_decoder(input); unicode_decoder) {
+    if (auto unicode_decoder = bom_sniff_to_decoder(input); unicode_decoder.has_value()) {
         // 1. Set encoding to BOMEncoding.
-        actual_decoder = unicode_decoder;
+        actual_decoder = &unicode_decoder.value();
 
         // 2. Read three bytes from ioQueue, if BOMEncoding is UTF-8; otherwise read two bytes. (Do nothing with those bytes.)
-        // FIXME: I imagine this will be pretty slow for large inputs, as it's regenerating the input without the first 2/3 bytes.
```
Comment (nico) [inline Userland/Libraries/LibTextCodec/Decoder.cpp:204]: Keep fix me?

**Review by nico — COMMENTED** [review]
Generally looks great (i stopped reading in the LibWeb part of the diff).

File: Userland/Libraries/LibTextCodec/Decoder.cpp (hunk 4 above)
Comment (AtkinsSJ) [reply Userland/Libraries/LibTextCodec/Decoder.cpp:204]: Why is getting a smaller StringView from a StringView a slow operation though?

File: Meta/Lagom/Fuzzers/FuzzCyrillicDecoder.cpp (hunk 2 above)
Comment (AtkinsSJ) [reply Meta/Lagom/Fuzzers/FuzzCyrillicDecoder.cpp:16]: That is a very good point! Thanks.

**Comment by AtkinsSJ** [comment]
Amended based on feedback. (Except the FIXME-removal.)

File: Userland/Libraries/LibTextCodec/Decoder.cpp (hunk 4 above)
Comment (nico) [reply Userland/Libraries/LibTextCodec/Decoder.cpp:204]: I think it tries to say that it makes a large copy when returning, not that this line here is slow (?)

(The general point is "don't make unrelated changes" – this PR isn't about this FIXME, so it shouldn't touch it imho.)

Comment (AtkinsSJ) [reply Userland/Libraries/LibTextCodec/Decoder.cpp:204]: I'll drop the commit, it seems silly to hold up the rest of the PR over this. :^)

**Comment by AtkinsSJ** [comment]
Dropped the commit that removed a FIXME. CI was all green before this so should be fine.

**Pull request merged**